### PR TITLE
Add ability to customize VolumeSnapshotContent workqueue

### PR DESCRIPTION
**Pull request by humblec**
Two new timeout values ( retryIntervalStart & retryIntervalMax )
have been added to set the ratelimiter for volumesnapshotcontent queue.

Fixes:  https://github.com/kubernetes-csi/external-snapshotter/issues/463

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

```release-note
 `retry-interval-start` and `retry-interval-max` arguments are added to csi-snapshotter sidecar which controls retry interval of failed volume snapshot creation and deletion. These values set the ratelimiter for volumesnapshotcontent queue.
```

### CODE REVIEW

**Comment by humblec** [comment]
@xing-yang @yuxiangqian ptal . thanks

**Comment by xing-yang** [comment]
@humblec Thanks for adding this. Requeue currently does not work properly so adding these options won't really do anything.  I'm working on a fix here: https://github.com/kubernetes-csi/external-snapshotter/pull/230.  After that we can get your fix in.

**Comment by humblec** [comment]
> @humblec Thanks for adding this. Requeue currently does not work properly so adding these options won't really do anything. I'm working on a fix here: #230. After that we can get your fix in.

@xing-yang sure. Will wait for other PR to go in first.

**Comment by xing-yang** [comment]
@humblec can you rebase?

**Comment by humblec** [comment]
> @humblec can you rebase?

@xing-yang  Sure, it is done. :+1:  ptal.

File: cmd/csi-snapshotter/main.go [hunk 1]
```diff
@@ -63,6 +61,9 @@ var (
 	threads                = flag.Int("worker-threads", 10, "Number of worker threads.")
 	csiTimeout             = flag.Duration("timeout", defaultCSITimeout, "The timeout for any RPCs to the CSI driver. Default is 1 minute.")
 
+	retryIntervalStart = flag.Duration("retry-interval-start", time.Second, "Initial retry interval of failed volumesnapshotcontent creation or deletion. It doubles with each failure, up to retry-interval-max.")
```
Comment (xing-yang) [inline cmd/csi-snapshotter/main.go:64]: s/volumesnapshotcontent/volume snapshot

File: cmd/csi-snapshotter/main.go [hunk 2]
```diff
@@ -63,6 +61,9 @@ var (
...
+	retryIntervalMax   = flag.Duration("retry-interval-max", 5*time.Minute, "Maximum retry interval of failed volumesnapshotcontent creation or deletion.")
```
Comment (xing-yang) [inline cmd/csi-snapshotter/main.go:65]: s/volumesnapshotcontent/volume snapshot

File: cmd/csi-snapshotter/main.go (hunk 1 above)
Comment (humblec) [reply cmd/csi-snapshotter/main.go]: Done.

File: cmd/csi-snapshotter/main.go (hunk 2 above)
Comment (humblec) [reply cmd/csi-snapshotter/main.go]: Done.

**Comment by humblec** [comment]
@xing-yang comments are addressed and also added release notes. Please let me know if any changes are required in any of it.

File: README.md [hunk 3]
```diff
@@ -26,10 +26,10 @@ Blog post for the beta feature can be found [here](https://kubernetes.io/blog/20
...
+| Compatible with CSI Version                                                                | Container Image                    | [Min K8s Version](https://kubernetes-csi.github.io/docs/kubernetes-compatibility.html#minimum-version) | Snapshot CRD version |
+| ------------------------------------------------------------------------------------------ | ---------------------------------- | ------------------------------------------------------------------------------------------------------ | -------------------- |
+| [CSI Spec v1.0.0](https://github.com/container-storage-interface/spec/releases/tag/v1.0.0) | quay.io/k8scsi/csi-snapshotter     | 1.17                                                                                                   | v1beta1              |
+| [CSI Spec v1.0.0](https://github.com/container-storage-interface/spec/releases/tag/v1.0.0) | quay.io/k8scsi/snapshot-controller | 1.17                                                                                                   | v1beta1              |
```
Comment (xing-yang) [inline README.md:32]: These changes are not related to this fix.

File: README.md [hunk 4]
```diff
@@ -90,9 +90,14 @@ Install CSI Driver:
 
 * `--metrics-path`: The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.
 
+* `--retry-interval-start`: Initial retry interval of failed volume snapshot creation or deletion. It doubles with each failure, up to retry-interval-max.
```
Comment (yuxiangqian) [inline README.md:93]: lets call it "retry-interval-base", _start_ sounds like a starting time. and lets add "<duration>" as resync-period.

File: README.md [hunk 5]
```diff
@@ -90,9 +90,14 @@ Install CSI Driver:
...
+
+* `--retry-interval-max`: Maximum retry interval of failed volume snapshot creation or deletion.
```
Comment (yuxiangqian) [inline README.md:95]: ditto

File: pkg/sidecar-controller/framework_test.go [hunk 6]
```diff
@@ -521,6 +522,7 @@ func newTestController(kubeClient kubernetes.Interface, clientset clientset.Inte
 		60*time.Second,
 		"snapshot",
 		-1,
+		workqueue.DefaultControllerRateLimiter(),
```
Comment (yuxiangqian) [inline pkg/sidecar-controller/framework_test.go:525]: why this is not changed to test the new rater limiter?

**Comment by yuxiangqian** [comment]
I am a little concerned for this change as it would change the behavior of the rateLimiter slightly.
1. how should this be effectively tested? I understand there are no existing tests to cover, but can we consider covering in this PR?
2. the original implementation uses the default controller rate limiter, which under the hood is just a combination of exponential failure rate limiter and a bucket rate limiter. For the exponential failure rate limiter, it uses 1 ms as the base. In this PR, the default base is 1s.
What are the main purposes of this change? is it just to allow users to specify the base and max-retry interval?

File: README.md (hunk 3 above)
Comment (humblec) [reply README.md:32]: Thats correct.

**Comment by humblec** [comment]
@yuxiangqian regarding the name of these parameters: While I agree that we could change or use some strings like `retry-interval-base` ..etc, we are actually deviating from the same parameters available in other sidecars and the parity between those. For ex: external provisioner/attacher (https://github.com/kubernetes-csi/external-attacher) too have these parameters.  Regarding the default value, we could change that to `1ms` or whatever preferred value , Is `1ms` good enough?

File: README.md (hunk 5 above)
Comment (humblec) [reply README.md:95]: https://github.com/kubernetes-csi/external-snapshotter/pull/308#issuecomment-634583568

File: README.md (hunk 4 above)
Comment (humblec) [reply README.md:93]: https://github.com/kubernetes-csi/external-snapshotter/pull/308#issuecomment-634583568

File: README.md (hunk 3 above)
Comment (xing-yang) [reply README.md:32]: Please remove these un-related changes.

**Comment by humblec** [comment]
> @yuxiangqian regarding the name of these parameters: While I agree that we could change or use some strings like `retry-interval-base` ..etc, we are actually deviating from the same parameters available in other sidecars and the parity between those. For ex: external provisioner/attacher (https://github.com/kubernetes-csi/external-attacher) too have these parameters. Regarding the default value, we could change that to `1ms` or whatever preferred value , Is `1ms` good enough?

@xing-yang can you share your thoughts.. I would like to get some consensus and hopefully finish this PR .

**Comment by xing-yang** [comment]
>> @yuxiangqian regarding the name of these parameters: While I agree that we could change or use some strings like retry-interval-base ..etc, we are actually deviating from the same parameters available in other sidecars and the parity between those. For ex: external provisioner/attacher (https://github.com/kubernetes-csi/external-attacher) too have these parameters. Regarding the default value, we could change that to 1ms or whatever preferred value , Is 1ms good enough?

> @xing-yang can you share your thoughts.. I would like to get some consensus and hopefully finish this PR .

I agree that the naming in externa-snapshotter should be consistent with that in other sidecars.  @yuxiangqian are you okay with this PR?

File: README.md (hunk 3 above)
Comment (xing-yang) [reply README.md:32]: Can you remove these changes from this PR?

**Comment by yuxiangqian** [comment]
> @yuxiangqian regarding the name of these parameters: While I agree that we could change or use some strings like `retry-interval-base` ..etc, we are actually deviating from the same parameters available in other sidecars and the parity between those. For ex: external provisioner/attacher (https://github.com/kubernetes-csi/external-attacher) too have these parameters. Regarding the default value, we could change that to `1ms` or whatever preferred value , Is `1ms` good enough?
>
@humblec dropped the ball a bit, my apologies. If its consistent with others, I am good with  the name.
I still have concerns over the default value of 1s as users likely would just rely on the default. Snapshot is latency sensitive and that's different from attacher. I'd  suggest add some tests to that and default to the original 1ms starting. and in  framework_test.go, should we also use the same ratelimiter as in main.go.

File: README.md (hunk 3 above)
Comment (humblec) [reply README.md:32]: Sure, removed.

Comment (humblec) [reply README.md]: @xing-yang removed in the latest patch set. PTAL.

**Comment by xing-yang** [comment]
@humblec can you reference this issue https://github.com/kubernetes-csi/external-snapshotter/issues/463 in the top section of this PR?

**Comment by humblec** [comment]
> @humblec can you reference this issue #463 in the top section of this PR?

Sure thing @xing-yang and thanks for reopening this PR, let me rebase and finish it off :+1:

**Comment by humblec** [comment]
The default value of the ratelimiter has been set to `1ms`.  Apart from that, one thing to note here is,  framework testing does not make use of this new rate limiter yet ie it use the defaultRateLimiter.  If required, I can change framework, otherwise we are good with the change .

File: README.md [hunk 7]
```diff
@@ -127,6 +127,9 @@ Read more about how to install the example webhook [here](deploy/kubernetes/webh
 
 * `--worker-threads`: Number of worker threads. Default value is 10.
 
+* `--retry-interval-start`: Initial retry interval of failed volume snapshot creation or deletion. It doubles with each failure, up to retry-interval-max.
```
Comment (xing-yang) [inline README.md:130]: Add default value.

File: README.md [hunk 8]
```diff
@@ -127,6 +127,9 @@ Read more about how to install the example webhook [here](deploy/kubernetes/webh
...
+
+*`--retry-interval-max`: Maximum retry interval of failed volume snapshot creation or deletion.
```
Comment (xing-yang) [inline README.md:132]: Add default value.

File: cmd/csi-snapshotter/main.go [hunk 9]
```diff
@@ -68,9 +69,11 @@ var (
...
-	metricsPath    = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+	metricsAddress     = flag.String("metrics-address", "", "(deprecated) The TCP network address where the prometheus metrics endpoint will listen (example: `:8080`). The default is empty string, which means metrics endpoint is disabled. Only one of `--metrics-address` and `--http-endpoint` can be set.")
+	httpEndpoint       = flag.String("http-endpoint", "", "The TCP network address where the HTTP server for diagnostics, including metrics and leader election health check, will listen (example: `:8080`). The default is empty string, which means the server is disabled. Only one of `--metrics-address` and `--http-endpoint` can be set.")
+	metricsPath        = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+	retryIntervalStart = flag.Duration("retry-interval-start", time.Millisecond, "Initial retry interval of failed volume snapshot creation or deletion. It doubles with each failure, up to retry-interval-max.")
```
Comment (xing-yang) [inline cmd/csi-snapshotter/main.go:75]: `time.Millisecond` is too short.  `time.Second` is fine.

File: pkg/sidecar-controller/framework_test.go (hunk 6 above)
Comment (xing-yang) [reply pkg/sidecar-controller/framework_test.go:526]: Can you address this comment?

File: README.md (hunk 8 above)
Comment (humblec) [reply README.md]: Sure, done

File: cmd/csi-snapshotter/main.go (hunk 9 above)
Comment (humblec) [reply cmd/csi-snapshotter/main.go]: Done

File: README.md (hunk 7 above)
Comment (humblec) [reply README.md]: Done

File: pkg/sidecar-controller/framework_test.go (hunk 6 above)
Comment (humblec) [reply pkg/sidecar-controller/framework_test.go]: done.. thanks ! @xing-yang ptal.

**Comment by humblec** [comment]
@xing-yang ptal.. Thanks !

File: pkg/sidecar-controller/framework_test.go [hunk 10]
```diff
@@ -522,6 +523,7 @@ func newTestController(kubeClient kubernetes.Interface, clientset clientset.Inte
 		"snapshot",
 		-1,
 		true,
+		workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 5*time.Minute),
```
Comment (xing-yang) [inline pkg/sidecar-controller/framework_test.go:526]: This is a test so it's better to set the time shorter.  Suggest changing to the following:
1*time.Second -> 1*time.Millisecond
5*time.Minute -> 1*time.Minute

Comment (humblec) [reply pkg/sidecar-controller/framework_test.go]: @xing-yang done. thanks.. ptal.
